services/batcher/batcher.go

```diff
@@ -0,0 +1,220 @@
+		tickerExit: make(chan bool, 1),
+
+		rh: fn,
+	}
+	bt.cache.SetDefault("nextCK", uint64(0))
+
+	/* handle ticker */
+	go func() {
+		for {
+			select {
+			case <-bt.ticker.C:
+				bt.batchBufferedMetrics()
+			case <-bt.tickerExit:
+				bt.batchBufferedMetrics()
+				return
+			}
+		}
+	}()
+
+	return &bt
+}
+
+// Add adds single transit.DynamicResourcesWithServicesRequest to batch
+func (bt *Batcher) Add(p []byte) error {
+	nextCK, err := bt.cache.IncrementUint64("nextCK", 1)
+	if err != nil {
+		return err
+	}
+	if nextCK == math.MaxUint64 {
+		nextCK = 0
+		bt.cache.SetDefault("nextCK", 0)
+	}
+	bt.cache.SetDefault(strconv.FormatUint(nextCK, 10), p)
+	bt.cacheSize += len(p)
+	if bt.cacheSize > bt.maxBytes {
+		log.Debug().Msgf("batch buffer size %dKB exceeded the soft limit %dKB",
+			bt.cacheSize/1024, bt.maxBytes/1024)
+		return bt.batchBufferedMetrics()
+	}
+	return err
+}
+
+// Exit stops the internal ticker
+func (bt *Batcher) Exit() {
+	bt.tickerExit <- true
+}
+
+// Reset applyes configuration
+func (bt *Batcher) Reset(d time.Duration, maxBytes int) {
+	bt.batchBufferedMetrics()
+	bt.maxBytes = maxBytes
+	if d == 0 {
+		d = math.MaxInt64
+	}
+	bt.ticker.Reset(d)
+}
+
+func (bt *Batcher) batchBufferedMetrics() error {
+	bld := NewBuilder()
+	for ck, c := range bt.cache.Items() {
+		if ck == "nextCK" {
+			continue
+		}
+		p := c.Object.([]byte)
+		bld.Add(p)
+		bt.cacheSize -= len(p)
+		bt.cache.Delete(ck)
+	}
+	payload, err := bld.Build()
+	if payload != nil && err == nil {
+		return bt.rh(context.Background(), payload)
+	}
+	return err
+}
+
+type sendRequestHandler func(context.Context, []byte) error
+
+// BatchBuilder implements builder
+type BatchBuilder struct {
+	tContexts []transit.TracerContext
+	groupsMap map[string]transit.ResourceGroup
+	resMap    map[string]transit.DynamicMonitoredResource
+	resSvcMap map[string][]transit.DynamicMonitoredService
+	svcMap    map[string]svcMapItem
+}
+
+// NewBuilder returns new instance
+func NewBuilder() *BatchBuilder {
+	return &BatchBuilder{
+		tContexts: make([]transit.TracerContext, 0),
+		groupsMap: make(map[string]transit.ResourceGroup),
+		resMap:    make(map[string]transit.DynamicMonitoredResource),
+		resSvcMap: make(map[string][]transit.DynamicMonitoredService),
+		svcMap:    make(map[string]svcMapItem),
+	}
+}
+
+// Add adds single transit.DynamicResourcesWithServicesRequest to batch
+func (bld *BatchBuilder) Add(p []byte) {
+	var r2 transit.DynamicResourcesWithServicesRequest
+	if err := json.Unmarshal(p, &r2); err != nil {
+		log.Err(err).
+			RawJSON("payload", p).
+			Msg("could not unmarshal metrics payload for batch")
+	} else {
+		bld.tContexts = append(bld.tContexts, *r2.Context)
+		for _, g := range r2.Groups {
+			bld.groupsMap[string(g.Type)+":"+g.GroupName] = g
+		}
+		for _, res := range r2.Resources {
+			resK := string(res.Type) + ":" + res.Name
+			for _, svc := range res.Services {
+				applyTime(&res, &svc, r2.Context.TimeStamp) // ensure time fields
```

Comment:
Not that I fully understand this chain of processing, but given the names of things. It appears there 'Groups' (on line 143) then there are 'Resources' which have 'Services'. 

Why don't the 'Resources' have 'Hosts' ? Or if 'Resources' are 'Hosts' then why are we only doing applyTime on the Services and not the parent Host?

Reply:
Please take a look on data structure definitions in the `transit` package. In general it reflects data structures of Foundation.
https://github.com/gwos/tcg/blob/master/transit/transit.go#L756-L761
And as you can see the `applyTime` processes both the Services and the Host.